src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -24,162 +25,252 @@ const (
+		epostMessage := _makeElectionPoStMessage(outTree, blk.Miner(), msgs.Epoch(), blk.PoStProof())
+		outTree = _applyMessageBuiltinAssert(outTree, epostMessage, blk.Miner())
+
+		minerPenaltyTotal := actor.TokenAmount(0)
+		var minerPenaltyCurr actor.TokenAmount
 
 		// Process BLS messages from the block.
 		for _, m := range blk.BLSMessages() {
 			_, found := seenMsgs[_msgCID(m)]
 			if found {
 				continue
 			}
-			outTree, r = vmi.ApplyMessage(outTree, m, blk.Miner())
-			receipts = append(receipts, r)
+			onChainMessageLen := len(msg.Serialize_UnsignedMessage(m))
+			outTree, receipt, minerPenaltyCurr = vmi.ApplyMessage(outTree, m, onChainMessageLen, blk.Miner())
+			minerPenaltyTotal += minerPenaltyCurr
+			receipts = append(receipts, receipt)
 			seenMsgs[_msgCID(m)] = struct{}{}
 		}
+
 		// Process SECP messages from the block.
 		for _, sm := range blk.SECPMessages() {
 			m := sm.Message()
 			_, found := seenMsgs[_msgCID(m)]
 			if found {
 				continue
 			}
-			outTree, r = vmi.ApplyMessage(outTree, m, blk.Miner())
-			receipts = append(receipts, r)
+			onChainMessageLen := len(msg.Serialize_SignedMessage(sm))
+			outTree, receipt, minerPenaltyCurr = vmi.ApplyMessage(outTree, m, onChainMessageLen, blk.Miner())
+			minerPenaltyTotal += minerPenaltyCurr
+			receipts = append(receipts, receipt)
 			seenMsgs[_msgCID(m)] = struct{}{}
 		}
+
+		deductAmount := minerPenaltyTotal
+		if deductAmount > outTree.Balance(minerOwner) {
+			deductAmount = outTree.Balance(minerOwner)
+		}
+		outTree = _withTransferFundsAssert(outTree, minerOwner, addr.BurntFundsActorAddr, deductAmount)
 	}
 
 	// Invoke cron tick, attributing it to the miner of the first block.
-	cronSender := msgs.Blocks()[0].Miner()
-	cron := _makeCronTickMessage(outTree, cronSender)
-	outTree, r = vmi.ApplyMessage(outTree, cron, cronSender)
-	if r.ExitCode() != exitcode.OK() {
-		panic("cron tick failed")
-	}
+
+	TODO()
+	// TODO: miners shouldn't be able to trigger cron by sending messages;
+	// use ControlActor instead (https://github.com/filecoin-project/specs/issues/665)
+
+	firstMiner := msgs.Blocks()[0].Miner()
+	cronMessage := _makeCronTickMessage(outTree, firstMiner)
+	outTree = _applyMessageBuiltinAssert(outTree, cronMessage, firstMiner)
 
 	return
 }
 
-func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.UnsignedMessage, minerAddr addr.Address) (
-	st.StateTree, vmr.MessageReceipt) {
+func (vmi *VMInterpreter_I) ApplyMessage(
+	inTree st.StateTree, message msg.UnsignedMessage, onChainMessageSize int, minerAddr addr.Address) (
+	retTree st.StateTree, retReceipt vmr.MessageReceipt, retMinerPenalty actor.TokenAmount) {
+
+	vmiGasRemaining := message.GasLimit()
 
-	compTree := inTree
-	var outTree st.StateTree
-	var toActor actor.ActorState
-	var err error
+	_applyReturn := func(
+		tree st.StateTree, invocOutput vmr.InvocOutput, exitCode exitcode.ExitCode,
+		gasUsedFrom *addr.Address, gasUsedTo *addr.Address, minerPenaltySpec MinerPenaltySpec) {
```

Comment:
Nit: This method is either called with final parameters `(..., nil, nil, MinerPenaltySpec_ChargeUsedGas)` or non-nils and `MinerPenaltySpec_ChargeUsedGas`, so it would be less noisy and easier to get right if there were separate methods and no penalty spec type.